### PR TITLE
Adding functionality to aggregate and annotate DeepProfiler output

### PR DESCRIPTION
**Pull request by gwaygenomics**
# Description

In this pull request, we complete the integration of DeepProfiler output and pycytominer processing.

We discuss the implementation in broadinstitute/DeepProfilerExperiments#2.

## What is the nature of your change?

- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

**Comment by gwaygenomics** [comment]
I added the real data tests, but the tests will fail until I add the data. The data will be added once https://github.com/broadinstitute/DeepProfilerExperiments/issues/2#issuecomment-698910369 is resolved.

**Comment by gwaygenomics** [comment]
@michaelbornholdt - one thing that you can add to this PR are [numpy style docstrings](https://www.sphinx-doc.org/en/master/usage/extensions/example_numpy.html) (which are compatible with Sphinx)

Example: https://github.com/cytomining/pycytominer/blob/fdd0e0935ab5446f5c3710e5270c5234d8ef18f1/pycytominer/aggregate.py#L26-L57

If you're looking deeply at this code, I suspect that you'll need to figure this out on a per-argument basis anyway. Thanks!

**Comment by gwaygenomics** [comment]
(once you accept the write access invitation, you'll be able to commit directly to this branch)

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (niranjchandrasekaran) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:29]: It is possible that this is a placeholder. If so, it should be changed.

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (niranjchandrasekaran) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:54]: I am not sure how the docstrings should look like for functions with no parameters and also don't return anything. Should it just be the following?

```
Returns
--------
None
        Nothing is returned.
```

If that is the case then both `setup_aggregate` and `aggregate_deep` should also have the same docstrings.

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (niranjchandrasekaran) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:62]: docstrings

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (niranjchandrasekaran) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:70]: docstrings

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (niranjchandrasekaran) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:147]: docstrings

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (niranjchandrasekaran) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:131]: Not sure if it is necessary but do we want to provide the functionality of aggregate only a subset of sites? If so, I guess we will have to add a filter to `setup_aggregate` and also change `__init__()` so that it can accept a list of fields, like in https://github.com/cytomining/pycytominer/blob/fdd0e0935ab5446f5c3710e5270c5234d8ef18f1/pycytominer/cyto_utils/cells.py#L52-L53 

We could also ignore this for the time being and add the functionality later.

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (niranjchandrasekaran) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:160]: I see that there is some metadata annotation happening in `load_npz()`. I don't think I have grasped the difference between what is happening there and what is happening in `annotate_deep()`. Can you explain what is different?

**Review by niranjchandrasekaran — CHANGES_REQUESTED** [review]
@gwaygenomics I have made some comments (mostly about docstrings) and have a couple of suggestions. Other than that everything looks good.

I also noticed that the [PR template](https://github.com/cytomining/pycytominer/blob/master/.github/PULL_REQUEST_TEMPLATE.md) was missing. I realize that this PR is old but can it still be added?

**Comment by gwaygenomics** [comment]
ooh, I see what happened now, I think once I marked the pull request as ready for review, the CODEOWNERS file triggered a review request:

> gwaygenomics requested a review from niranjchandrasekaran as a code owner 5 days ago

In the future and to save Niranj time, I will not be so quick to mark a PR as ready-for-review

**Comment by gwaygenomics** [comment]
> I also noticed that the PR template was missing. I realize that this PR is old but can it still be added?

Yeah, it's an old PR - but I agree adding it would be helpful. @michaelbornholdt, when you are happy with your fixes, please go through the checklist [above](https://github.com/cytomining/pycytominer/pull/78#issue-415389650).

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (michaelbornholdt) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:127]: This part is giving me Errors @gwaygenomics. 
The features list is a list of all columns (Metadata and profiles) this gives an error in aggregate because it expects only the profiles, I think. 
If you don't know the solution, I will work on it

Comment (gwaygenomics) [reply pycytominer/cyto_utils/DeepProfiler_processing.py:127]: what is the error?

at this point, df should just be features, plus a column Metadata_Aggregate_On - but the features should not include this variable. (the metadata should all be in metadata_df)

File: pycytominer/cyto_utils/load.py
Comment (michaelbornholdt) [reply pycytominer/cyto_utils/load.py]: Done

File: pycytominer/cyto_utils/load.py
Comment (michaelbornholdt) [reply pycytominer/cyto_utils/load.py:84]: done

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (michaelbornholdt) [reply pycytominer/cyto_utils/DeepProfiler_processing.py:55]: deleted it

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (michaelbornholdt) [reply pycytominer/cyto_utils/DeepProfiler_processing.py]: I will add to the list of additial things, not in this PR

Comment (michaelbornholdt) [reply pycytominer/cyto_utils/DeepProfiler_processing.py]: #144

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (michaelbornholdt) [reply pycytominer/cyto_utils/DeepProfiler_processing.py]: See discussion below on inconsistencies

**Comment by michaelbornholdt** [comment]
@gwaygenomics 
Done. Don't know what that check is and why its failing.

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (gwaygenomics) [inline pycytominer/cyto_utils/DeepProfiler_processing.py:206]: I am a bit concerned about this line. The function will error out if the info in Metadata_Site cannot be coerced to int. Is this line necessary?

Comment (michaelbornholdt) [reply pycytominer/cyto_utils/DeepProfiler_processing.py:206]: Yea, its part of the spageti code of this function. 

the index files has the sites and ints and I think they should be ints. 
But when you read them from the npz files, then you get str... so one of them has to be changed

Comment (gwaygenomics) [reply pycytominer/cyto_utils/DeepProfiler_processing.py:206]: but why do you need them to be ints? You can groupby str variables and I think that's all we need this info for (aggregation)

Comment (gwaygenomics) [reply pycytominer/cyto_utils/DeepProfiler_processing.py]: In 353cb382db3783a9febaca8fc02c96598f2bebfd I made sure that all metadata columns have an `str` dtype upon instantiation.

I think this is our safest bet.

**Comment by gwaygenomics** [comment]
@niranjchandrasekaran - this is back to you. Thanks!

File: pycytominer/cyto_utils/DeepProfiler_processing.py
Comment (michaelbornholdt) [reply pycytominer/cyto_utils/DeepProfiler_processing.py]: OK sure, checks out now

**Review by niranjchandrasekaran — APPROVED** [review]
@gwaygenomics @michaelbornholdt LGTM! You may want to update the docstrings for the old functions in `pycytominer/cyto_utils/load.py` while you are at it. But it can also wait. Nevertheless I have approved the PR.

**Comment by gwaygenomics** [comment]
awesome, thanks @niranjchandrasekaran ! 

> You may want to update the docstrings for the old functions in pycytominer/cyto_utils/load.py

I definitely think we should update docstrings as we go - but I think we did update these functions? Maybe I'm missing something or you mean a different file?

Either way, that potential documentation update is relatively minor - which is great!

@michaelbornholdt is this ready to merge? If so, I will go ahead.

**Comment by niranjchandrasekaran** [comment]
> I definitely think we should update docstrings as we go - but I think we did update these functions? Maybe I'm missing something or you mean a different file?

Ah, it was just a single function.

https://github.com/cytomining/pycytominer/blob/d2db94733ca1c9e538517f0f728866886ccc7e41/pycytominer/cyto_utils/load.py#L7-L13

**Comment by michaelbornholdt** [comment]
> @michaelbornholdt is this ready to merge? If so, I will go ahead.

All good! Go ahead. 
I already have the next PR set up in my head 🤣

**Comment by gwaygenomics** [comment]
> Ah, it was just a single function.

ooo, good catch. Thanks. I will update.

> All good! Go ahead.
I already have the next PR set up in my head 🤣

Wonderful. Can't wait! 😈

**Pull request merged**
